File: typescript/packages/mechanisms/evm/src/batch-settlement/storage/query.ts
import type { Network } from "@x402/core/types";
import type { Channel, ChannelStorage } from "./channel";

/** Per-call store options. */
export type ChannelStoreOptions = { signal?: AbortSignal };

/**
 * Closed set of worker reads the channel managers issue.
 *
 * A named union rather than optional predicates: an adapter that does not
 * understand a field would otherwise over-return, and over-returning is the
 * dangerous direction (cooperative refund of channels that are not idle).
 * Switching on `kind` with a `never` default makes a new variant a compile
 * error in every custom adapter.
 */
export type ChannelQuery =
  | {
      kind: "claimable";
      network?: Network;
      idleAtOrBefore?: number;
      limit?: number;
      cursor?: string;
    }
  | {
      kind: "idleRefundable";
      network?: Network;
      idleAtOrBefore?: number;
      limit?: number;
      cursor?: string;
    }
  | { kind: "withdrawPending"; network?: Network; limit?: number; cursor?: string };

/**
 * Filter for {@link ChannelStorage.settleQuery}: distinct claimed
 * `(network, receiver, token)` tuples.
 */
export type SettleQuery = {
  network?: Network;
  limit?: number;
  cursor?: string;
};

/** Distinct claimed `(network, receiver, token)` used by facilitator settle. */
export type SettleTarget = [network: Network, receiver: `0x${string}`, token: `0x${string}`];

/** One page from {@link ChannelStorage.query} or {@link ChannelStorage.settleQuery}. */
export type QueryPage<T> = { items: T[]; cursor?: string };

/**
 * Returns whether `channel` satisfies `filter`.
 *
 * Uint256 comparisons use `BigInt` so lexicographic string order cannot leak
 * into adapters. When `filter.network` is set, a row without a `network` field
 * does not match.
 *
 * @param channel - Stored channel record.
 * @param filter - Named worker query.
 * @returns Whether the row belongs in the result set.
 */
export function matchesChannelQuery(channel: Channel, filter: ChannelQuery): boolean {
  if (filter.network !== undefined && channelNetwork(channel) !== filter.network) {
    return false;
  }

  switch (filter.kind) {
    case "claimable":
      if (BigInt(channel.chargedCumulativeAmount) <= BigInt(channel.totalClaimed)) {
        return false;
      }
      return matchesIdle(channel, filter.idleAtOrBefore);
    case "idleRefundable":
      if (BigInt(channel.balance) === 0n) {
        return false;
      }
      return matchesIdle(channel, filter.idleAtOrBefore);
    case "withdrawPending":
      return channel.withdrawRequestedAt > 0;
    default: {
      const _exhaustive: never = filter;
      throw new Error(`unhandled channel query: ${String(_exhaustive)}`);
    }
  }
}

/**
 * Orders query matches. Claimable rows put withdraw-pending channels first;
 * other kinds preserve input order.
 *
 * @param channels - Rows already accepted by {@link matchesChannelQuery}.
 * @param filter - Named worker query.
 * @returns A new array in worker-consumption order.
 */
export function sortChannels<T extends Channel>(channels: T[], filter: ChannelQuery): T[] {
  switch (filter.kind) {
    case "claimable":
      return [...channels].sort((a, b) => {
        const pendingA = a.withdrawRequestedAt > 0 ? 0 : 1;
        const pendingB = b.withdrawRequestedAt > 0 ? 0 : 1;
        return pendingA - pendingB;
      });
    case "idleRefundable":
    case "withdrawPending":
      return [...channels];
    default: {
      const _exhaustive: never = filter;
      throw new Error(`unhandled channel query: ${String(_exhaustive)}`);
    }
  }
}

/**
 * Scan-backed {@link ChannelStorage.query}. Dumps `list()`, filters, sorts, and
 * slices. `limit` / `cursor` only page the already-loaded match list; a native
 * adapter should apply them as read bounds.
 *
 * @param storage - Channel store to scan.
 * @param filter - Named worker query.
 * @returns One page of matching channels.
 */
export async function queryByScan<T extends Channel>(
  storage: ChannelStorage<T>,
  filter: ChannelQuery,
): Promise<QueryPage<T>> {
  const matched = sortChannels(
    (await storage.list()).filter(channel => matchesChannelQuery(channel, filter)),
    filter,
  );
  return pageItems(matched, filter.limit, filter.cursor);
}

/**
 * Scan-backed {@link ChannelStorage.settleQuery}: claimed rows
 * (`totalClaimed > 0`) deduped per `(network, receiver, token)`.
 *
 * @param storage - Channel store to scan.
 * @param filter - Settle query.
 * @returns One page of distinct settle targets in first-seen order.
 */
export async function settleQueryByScan<T extends Channel>(
  storage: ChannelStorage<T>,
  filter: SettleQuery = {},
): Promise<QueryPage<SettleTarget>> {
  const targets = new Map<string, SettleTarget>();
  for (const channel of await storage.list()) {
    if (BigInt(channel.totalClaimed) === 0n) {
      continue;
    }
    const network = channelNetwork(channel) ?? filter.network;
    if (!network || (filter.network !== undefined && network !== filter.network)) {
      continue;
    }
    const receiver = channel.channelConfig.receiver;
    const token = channel.channelConfig.token;
    const key = `${network}:${receiver.toLowerCase()}:${token.toLowerCase()}`;
    if (!targets.has(key)) {
      targets.set(key, [network, receiver, token]);
    }
  }
  return pageItems([...targets.values()], filter.limit, filter.cursor);
}

/**
 * Runs a named worker query, using a native `storage.query` when present and
 * {@link queryByScan} otherwise.
 *
 * @param storage - Channel store.
 * @param filter - Named worker query.
 * @param opts - Optional per-call store options.
 * @returns One page of matching channels.
 */
export async function queryChannels<T extends Channel>(
  storage: ChannelStorage<T>,
  filter: ChannelQuery,
  opts?: ChannelStoreOptions,
): Promise<QueryPage<T>> {
  return (await storage.query?.(filter, opts)) ?? queryByScan(storage, filter);
}

/**
 * Lists distinct claimed settle targets, using native `storage.settleQuery`
 * when present and {@link settleQueryByScan} otherwise.
 *
 * @param storage - Channel store.
 * @param filter - Settle query.
 * @param opts - Optional per-call store options.
 * @returns One page of distinct settle targets.
 */
export async function querySettleTargets<T extends Channel>(
  storage: ChannelStorage<T>,
  filter: SettleQuery = {},
  opts?: ChannelStoreOptions,
): Promise<QueryPage<SettleTarget>> {
  return (await storage.settleQuery?.(filter, opts)) ?? settleQueryByScan(storage, filter);
}

/**
 * Returns whether `channel` is idle at or before `idleAtOrBefore`.
 *
 * @param channel - Stored channel record.
 * @param idleAtOrBefore - Inclusive last-request cutoff; omitted means no idle filter.
 * @returns Whether the idle predicate passes.
 */
function matchesIdle(channel: Channel, idleAtOrBefore?: number): boolean {
  if (idleAtOrBefore === undefined) {
    return true;
  }
  return channel.lastRequestTimestamp <= idleAtOrBefore;
}

/**
 * Reads a CAIP-2 network from a row when the record carries one.
 *
 * @param channel - Stored channel record.
 * @returns Network when present on the row.
 */
function channelNetwork(channel: Channel): Network | undefined {
  const network = (channel as Channel & { network?: unknown }).network;
  return typeof network === "string" ? (network as Network) : undefined;
}

/**
 * Slices a match list using the scan-shim cursor (decimal offset).
 *
 * @param items - Sorted matches.
 * @param limit - Optional page size.
 * @param cursor - Opaque offset from a previous shim page.
 * @returns One page and the next cursor when more rows remain.
 */
function pageItems<T>(items: T[], limit?: number, cursor?: string): QueryPage<T> {
  const start = parseQueryCursor(cursor);
  const remaining = items.length - start;
  if (remaining <= 0) {
    return { items: [] };
  }
  const size = limit === undefined ? remaining : Math.max(0, limit);
  const end = start + size;
  const page = items.slice(start, end);
  return {
    items: page,
    ...(end < items.length ? { cursor: String(end) } : {}),
  };
}

/**
 * Parses a scan-shim cursor as a non-negative decimal offset.
 *
 * @param cursor - Opaque cursor from a previous page.
 * @returns Start index into the sorted match list.
 */
function parseQueryCursor(cursor: string | undefined): number {
  if (cursor === undefined || cursor === "") {
    return 0;
  }
  const parsed = Number.parseInt(cursor, 10);
  if (!Number.isFinite(parsed) || parsed < 0) {
    return 0;
  }
  return parsed;
}

File: go/mechanisms/evm/batch-settlement/storage/query.go
package storage

import (
	"context"
	"math/big"
	"strconv"
	"strings"
)

// ChannelStoreOptions are per-call options for indexed queries.
type ChannelStoreOptions struct {
	// Unused by the scan shim; native adapters may honour cancellation.
}

// QueryKind is the closed set of worker reads channel managers issue.
type QueryKind string

const (
	QueryKindClaimable       QueryKind = "claimable"
	QueryKindIdleRefundable  QueryKind = "idleRefundable"
	QueryKindWithdrawPending QueryKind = "withdrawPending"
)

// ChannelQuery is a named worker query. Adapters that ignore a field would
// over-return, and over-returning is the dangerous direction (cooperative
// refund of channels that are not idle).
type ChannelQuery struct {
	Kind           QueryKind
	Network        string
	IdleAtOrBefore *int64
	// MinUnclaimed is an optional decimal uint256 unclaimed threshold; absent means any positive unclaimed.
	MinUnclaimed *string
	// UnclaimedDesc sorts claimable rows highest-unclaimed first.
	UnclaimedDesc bool
	// OldestFirst sorts claimable rows by lastRequestTimestamp ascending (withdraw-pending first, then earliest withdrawRequestedAt).
	OldestFirst bool
	Limit       *int
	Cursor      string
}

// SettleQuery filters distinct claimed (network, receiver, token) tuples.
type SettleQuery struct {
	Network    string
	Limit      *int
	Cursor     string
	MinPending *big.Int
}

// SettleTarget is a distinct claimed (network, receiver, token) used by facilitator settle.
type SettleTarget struct {
	Network  string
	Receiver string
	Token    string
}

// QueryPage is one page from Query or SettleQuery.
type QueryPage[T any] struct {
	Items  []T
	Cursor string
}

// ChannelQuerier is an optional indexed worker query. QueryChannels type-asserts
// this and falls back to QueryByScan when it is absent.
type ChannelQuerier[T ChannelRecord[T]] interface {
	Query(ctx context.Context, filter ChannelQuery, opts *ChannelStoreOptions) (*QueryPage[T], error)
}

// ChannelReceiverTokenQuerier lists channels for one receiver and token.
type ChannelReceiverTokenQuerier[T ChannelRecord[T]] interface {
	QueryByReceiverToken(ctx context.Context, network, receiver, token string) ([]T, error)
}

// MatchesChannelQuery reports whether channel satisfies filter.
//
// Uint256 comparisons use big.Int so lexicographic string order cannot leak
// into adapters. When filter.Network is set, a row without a Network field
// does not match.
func MatchesChannelQuery(channel *Channel, filter ChannelQuery) bool {
	if channel == nil {
		return false
	}
	if filter.Network != "" && channelNetwork(channel) != filter.Network {
		return false
	}

	switch filter.Kind {
	case QueryKindClaimable:
		charged, chargedOk := ParseUint256(channel.ChargedCumulativeAmount)
		claimed, claimedOk := ParseUint256(channel.TotalClaimed)
		if !chargedOk || !claimedOk || charged.Cmp(claimed) <= 0 {
			return false
		}
		if filter.MinUnclaimed != nil && filter.IdleAtOrBefore != nil {
			return matchesUnclaimedThreshold(charged, claimed, *filter.MinUnclaimed) ||
				matchesIdle(channel, filter.IdleAtOrBefore)
		}
		if filter.MinUnclaimed != nil {
			return matchesUnclaimedThreshold(charged, claimed, *filter.MinUnclaimed)
		}
		return matchesIdle(channel, filter.IdleAtOrBefore)
	case QueryKindIdleRefundable:
		balance, ok := ParseUint256(channel.Balance)
		if !ok || balance.Sign() == 0 {
			return false
		}
		return matchesIdle(channel, filter.IdleAtOrBefore)
	case QueryKindWithdrawPending:
		return channel.WithdrawRequestedAt > 0
	default:
		panic("unhandled channel query: " + string(filter.Kind))
	}
}

// SortChannels orders query matches. Claimable rows put withdraw-pending
// first, then highest-unclaimed when UnclaimedDesc is set, or oldest
// lastRequestTimestamp when OldestFirst is set.
func SortChannels[T ChannelRecord[T]](channels []T, filter ChannelQuery) []T {
	out := append([]T(nil), channels...)
	switch filter.Kind {
	case QueryKindClaimable:
		stableSortChannels(out, func(a, b T) bool {
			pendingA := 1
			if a.Base().WithdrawRequestedAt > 0 {
				pendingA = 0
			}
			pendingB := 1
			if b.Base().WithdrawRequestedAt > 0 {
				pendingB = 0
			}
			if pendingA != pendingB {
				return pendingA < pendingB
			}
			if filter.UnclaimedDesc {
				return unclaimedValue(a.Base()).Cmp(unclaimedValue(b.Base())) > 0
			}
			if filter.OldestFirst {
				baseA := a.Base()
				baseB := b.Base()
				if baseA.WithdrawRequestedAt > 0 && baseB.WithdrawRequestedAt > 0 &&
					baseA.WithdrawRequestedAt != baseB.WithdrawRequestedAt {
					return baseA.WithdrawRequestedAt < baseB.WithdrawRequestedAt
				}
				return baseA.LastRequestTimestamp < baseB.LastRequestTimestamp
			}
			return false
		})
		return out
	case QueryKindIdleRefundable, QueryKindWithdrawPending:
		return out
	default:
		panic("unhandled channel query: " + string(filter.Kind))
	}
}

// QueryByScan dumps List(), filters, sorts, and slices. Limit/cursor only page
// the already-loaded match list; a native adapter should apply them as read bounds.
func QueryByScan[T ChannelRecord[T]](ctx context.Context, store ChannelStorage[T], filter ChannelQuery) (*QueryPage[T], error) {
	all, err := store.List(ctx)
	if err != nil {
		return nil, err
	}
	matched := make([]T, 0, len(all))
	for _, channel := range all {
		if MatchesChannelQuery(channel.Base(), filter) {
			matched = append(matched, channel)
		}
	}
	return PageItems(SortChannels(matched, filter), filter.Limit, filter.Cursor), nil
}

// QueryChannels runs a named worker query, using a native ChannelQuerier when
// present and QueryByScan otherwise.
func QueryChannels[T ChannelRecord[T]](ctx context.Context, store ChannelStorage[T], filter ChannelQuery, opts *ChannelStoreOptions) (*QueryPage[T], error) {
	if querier, ok := store.(ChannelQuerier[T]); ok {
		page, err := querier.Query(ctx, filter, opts)
		if err != nil {
			return nil, err
		}
		if page != nil {
			return page, nil
		}
	}
	return QueryByScan(ctx, store, filter)
}

// QueryChannelsByReceiverToken loads channels for cleanup after settle.
func QueryChannelsByReceiverToken[T ChannelRecord[T]](
	ctx context.Context,
	store ChannelStorage[T],
	network, receiver, token string,
) ([]T, error) {
	if querier, ok := store.(ChannelReceiverTokenQuerier[T]); ok {
		return querier.QueryByReceiverToken(ctx, network, receiver, token)
	}
	all, err := store.List(ctx)
	if err != nil {
		return nil, err
	}
	out := make([]T, 0)
	wantReceiver := strings.ToLower(receiver)
	wantToken := strings.ToLower(token)
	for _, row := range all {
		base := row.Base()
		if base == nil {
			continue
		}
		if network != "" && channelNetwork(base) != network {
			continue
		}
		if strings.ToLower(base.ChannelConfig.Receiver) != wantReceiver ||
			strings.ToLower(base.ChannelConfig.Token) != wantToken {
			continue
		}
		out = append(out, row)
	}
	return out, nil
}

func matchesIdle(channel *Channel, idleAtOrBefore *int64) bool {
	if idleAtOrBefore == nil {
		return true
	}
	return channel.LastRequestTimestamp <= *idleAtOrBefore
}

// matchesUnclaimedThreshold reports whether charged-claimed meets threshold.
// An unparseable threshold fails closed.
func matchesUnclaimedThreshold(charged, claimed *big.Int, minUnclaimed string) bool {
	threshold, ok := ParseUint256(minUnclaimed)
	if !ok {
		return false
	}
	return new(big.Int).Sub(charged, claimed).Cmp(threshold) >= 0
}

// unclaimedValue returns charged-claimed, or zero when corrupt. Ordering
// helper for already-matched rows only.
func unclaimedValue(channel *Channel) *big.Int {
	charged, chargedOk := ParseUint256(channel.ChargedCumulativeAmount)
	claimed, claimedOk := ParseUint256(channel.TotalClaimed)
	if !chargedOk || !claimedOk {
		return new(big.Int)
	}
	delta := new(big.Int).Sub(charged, claimed)
	if delta.Sign() < 0 {
		return new(big.Int)
	}
	return delta
}

func channelNetwork(channel *Channel) string {
	if channel == nil {
		return ""
	}
	return channel.Network
}

func PageItems[T any](items []T, limit *int, cursor string) *QueryPage[T] {
	start := ParseQueryCursor(cursor)
	remaining := len(items) - start
	if remaining <= 0 {
		return &QueryPage[T]{Items: []T{}}
	}
	size := remaining
	if limit != nil {
		size = *limit
		if size < 0 {
			size = 0
		}
	}
	end := start + size
	if end > len(items) {
		end = len(items)
	}
	page := append([]T(nil), items[start:end]...)
	out := &QueryPage[T]{Items: page}
	if end < len(items) {
		out.Cursor = strconv.Itoa(end)
	}
	return out
}

func ParseQueryCursor(cursor string) int {
	if cursor == "" {
		return 0
	}
	parsed, err := strconv.Atoi(cursor)
	if err != nil || parsed < 0 {
		return 0
	}
	return parsed
}

// ParseUint256 parses a decimal uint256. ok=false means the caller must skip
// (query/claim scans) or fail closed (writes, verify) — never treat as zero.
func ParseUint256(s string) (*big.Int, bool) {
	v, ok := new(big.Int).SetString(s, 10)
	if !ok || v.Sign() < 0 {
		return nil, false
	}
	return v, true
}

// Uint256Cmp compares decimal uint256 strings. ok=false means either operand
// failed to parse, and the caller must skip the row, not treat it as zero.
func Uint256Cmp(a, b string) (int, bool) {
	ai, okA := ParseUint256(a)
	if !okA {
		return 0, false
	}
	bi, okB := ParseUint256(b)
	if !okB {
		return 0, false
	}
	return ai.Cmp(bi), true
}

func stableSortChannels[T ChannelRecord[T]](channels []T, less func(a, b T) bool) {
	// Insertion sort keeps equal-priority rows in input order.
	for i := 1; i < len(channels); i++ {
		j := i
		for j > 0 && less(channels[j], channels[j-1]) {
			channels[j], channels[j-1] = channels[j-1], channels[j]
			j--
		}
	}
}
